### PR TITLE
SubEns: surface hopping

### PR DESCRIPTION
**Pull request by tbereau**
- Fixes tabulated dihedral potential (angle was wrong; force propagation was wrong)
- Implements surface hopping MD. Method presented in:

https://arxiv.org/abs/1808.05644

Surface hopping introduces structural correlations between interaction terms. Currently implemented for bond, bending angle, and dihedral.

### CODE REVIEW

**Comment by junghans** [comment]
Something is wrong here, e.g. 1c3c8839fc4495e3f866cd4d145a78e7a3f9876e is the same as 6478c59d538b3d52fac0661432f6f68b89a3010a in the master branch, did you rebase or something?

**Comment by tbereau** [comment]
No I just did `git pull upstream master` from `espressopp/espressopp`. Indeed it seems like there are several recent commits that show up twice. I think it's because it `upstream`ed but then github tried to outsmart me and offered to update the branch. Must have applied the same update twice. I guess I need to revert to an older commit?

**Comment by tbereau** [comment]
Should be good now. I did a `reset` before my merge to the `upstream` and there doesn't seem to be any duplicate anymore.

**Comment by junghans** [comment]
I guess, a bunch of tests would be nice.

**Comment by tbereau** [comment]
Is there some documentation somewhere on how to add tests? Can't find anything in the manual or website.

File: src/interaction/TabulatedSubEns.cpp [hunk 1]
```diff
@@ -0,0 +1,318 @@
+/*
+  Copyright (C) 2016
```
Comment (jkrajniak) [inline src/interaction/TabulatedSubEns.cpp]: please update copyright

File: src/interaction/AngularPotential.py [hunk 2]
```diff
@@ -28,20 +28,21 @@
...
-		:rtype: 
+		:param \*args:
+		:type \*args:
+		:rtype:
 """
 # -*- coding: iso-8859-1 -*-
 from espressopp import pmi
 from espressopp import toReal3DFromVector
 from _espressopp import interaction_AngularPotential
+from espressopp import esutil
```
[the remark this answers is not in the document]
Comment (tbereau) [reply src/interaction/AngularPotential.py]: It's a mistake, thanks for pointing it out. I've now removed it.

File: src/interaction/TabulatedSubEns.cpp (hunk 1 above)
Comment (tbereau) [reply src/interaction/TabulatedSubEns.cpp]: Done.

**Comment by tbereau** [comment]
Thanks, @jkrajniak. I've added some comments about the `0.98` value. I've pushed some updates. `travis-ci` is running.

File: src/RealND.hpp [hunk 3]
```diff
@@ -110,13 +112,51 @@ namespace espressopp {
...
+      dimension = _dim;
+    }
+    int getDimension() const { return dimension; }
+
+    RealNDs();
+
+    // assignment is not the same as initialization
+    RealNDs& operator=(const RealNDs& v);
+
+    RealND& operator[](int i);
```
[the remark this answers is not in the document]
Comment (tbereau) [reply src/RealND.hpp]: No, these are operators for `RealNDs` that return a `RealND` :)

File: src/interaction/DihedralPotential.cpp [hunk 4]
```diff
@@ -27,6 +27,37 @@
...
+    DihedralPotential::computePhi(const Real3D& r21,
+                 const Real3D& r32,
+                 const Real3D& r43) {
+      Real3D rijjk = r21.cross(r32); // [r21 x r32]
+      Real3D rjkkn = r32.cross(r43); // [r32 x r43]
+
+      real rijjk_sqr = rijjk.sqr();
+      real rjkkn_sqr = rjkkn.sqr();
+
+      real rijjk_abs = sqrt(rijjk_sqr);
```
[the remark this answers is not in the document]
Comment (tbereau) [reply src/interaction/DihedralPotential.cpp]: You're right. I'll change that.

**Comment by tbereau** [comment]
@niktre Thanks. I've pushed the revised code.

**Pull request merged**